### PR TITLE
Thanasis/184 setup firebase google sign in

### LINKED ISSUE #184: [Fullstack] Setup Firebase Google Sign In
## Context

BEI wants a way to monitor patient progress and corresponding analytics, such as the number of sessions completed and time spent per question. For this, we want to add support for patients to sign in easily using a Google provider.

## Requirements

Setup Firebase client authentication in the BEI app, and update `components/Auth/LoginButton/LoginButton.tsx` and `components/Auth/LogoutButton.tsx` with appropriate sign-in and sign-out functions. If users have previously signed in and haven't cleared their caches, load the auth information from the SDK's cache (or any appropriate methods).

The app should attach an authenticated user's idToken to every request sent to our server (https://github.com/GTBitsOfGood/brain-exercise-analytics). Edit our server's APIWrapper (https://github.com/GTBitsOfGood/brain-exercise-analytics/blob/main/server/utils/APIWrapper.ts) to extract the email from the idToken for usage in server routes. 

Finally, we shall set up the signup process. The user flow should be as follows:
- If user is not signed in, `AuthGuard.tsx` should display the `LoginButton`
- On clicking the `LoginButton`, a request should be made to `/api/patient/auth/login` (description here: https://github.com/GTBitsOfGood/brain-exercise-analytics/issues/2)
- The route will return the user info. Save this state in `authReducer`.
- Build a SignUpGuard component based on `SignUp/Option1.tsx` which returns the main app if authReducer's `user.signedUp` is `true` else return the `SignUp` page.
- Once the user finishes filling out their sign-up details, send a request to `/api/patient/auth/signup` with the appropriate info from the page and update `authReducer` with the response data.
- Since this route should set `user.signedUp` to true, the `SignUpGuard` will automatically return the main app.

## Acceptance Criteria

- [x] Setup Firebase for the BEI app
- [x] Update sign-in and sign-out functions (the latter should also clear authReducer's state)
- [x] Load firebase state if previously logged in
- [ ] Make appropriate changes in the server's Firebase admin and APIWrapper to authenticate idTokens attached to requests, and decode the emails from them.
- [x] Add user info to `authReducer` after signing in with the `SignInButton` and sending a request to our server's `/api/patient/auth/login` route
- [x] If the user has not signed up, they should be taken to the SignUp page where they can fill in their details and send a request to `/api/patient/auth/signup` route. The `authReducer` should again be updated
- [x] Only if the user is authenticated and signed up, they can access the main app.

### PR DESCRIPTION
**Pull request by thanasis457**
## PR Title

Issue Number(s): #184 .

What does this PR change and why?
It introduces proper sign in with Firebase, caching of credentials locally, and updates the Analytics backend.
Also brings Auth Guard to prevent access to non-authenticated users.
Lastly, many small but important changes have happened (and added some comments here and there for future tickets).

### Checklist
- [x] Setup Firebase for the BEI app
- [x] Update sign-in and sign-out functions (the latter should also clear authReducer's state)
- [x] Load firebase state if previously logged in
- [ ] Make appropriate changes in the server's Firebase admin and APIWrapper to authenticate idTokens attached to requests, and decode the emails from them.
- [x] Add user info to `authReducer` after signing in with the `SignInButton` and sending a request to our server's `/api/patient/auth/login` route
- [x] If the user has not signed up, they should be taken to the SignUp page where they can fill in their details and send a request to `/api/patient/auth/signup` route. The `authReducer` should again be updated
- [x] Only if the user is authenticated and signed up, they can access the main app.

### Critical Changes / Notes
An important note is that the Analytics endpoints that should be updated on login have not been implemented yet. The code is there but untested and commented out.

#### Additional Notes
Here are some important updates and considerations for the codebase:
- OAuth0 is being used but sparsely and vaguely throughout the codebase. Not sure what it's use since we have firebase handling the authentication. (Edit: Auth0 is now removed)
- The UI is very buggy on the login and sign up screens (edit: The UI is now fixed)
- Google Sign In is very difficult to implement because the redirect uri is a local address during development. All React Native/Firebase integration libraries require ejection from Expo Go.
- We need to make a robust outline for how information flows (both for the sign in and for other data). For now the way sign in is handled is the following: During sign in or sign up, the authreducer gets called and it is the one for handling both state changes and async storage saves. Simultaneously, the firebase auth instance is logged in to the user.
During subsequent logins, firebase logins automatically from persistent storage and and retrieves the information (with async storage) and updates the reducer.
- The behaviour for the codebase is undefined if during sign up the person does not submit his Personal Info or if the Analytics Backend fails. It would be best to catch such errors and erase the user from the firebase db too.
The Auth Guard should block the user if he hasn't submitted his Personal Info but since the endpoint has not been implemented yet it will be added later.
- The personal info does not yet pop up (pending endpoint implementation).
- The axios instance on the mobile app attached the id of the user. The backend code (in analytics) is there, commented out. Not sure how the wrapper works so might get some help from someone who understands it a bit better.

Problems:
1. What if firebase gets logged out while the user profile is still cached?
2. What if the cached user info gets lost while firebase is still authenticated?
3. What happens offline?

Solutions:
1. The Auth Guard does not depend on the local user info, but on the Firebase Auth instance. This ensures that a user is truly logged in when Firebase tells us so. We are not dependent on fragile cache data.
2. Not implemented yet but will probably just add code to refresh the local user data alongside the firebase authentication process.
This makes Async storage useless, however, since we will be updating the local info (by querying analytics) pretty much every time, making caching useless.
3. The only use for the Async storage is to retrieve user information when we cannot log in to Firebase due to connection errors. Implementation pending.

Additional:
If we need to make more complex sign in methods or handling corner cases it would be good to abstract this logic to other components since right now most of the code is "in-place".

### Testing
Make a new account and sign in.

### CODE REVIEW

**Comment by thanasis457** [comment]
Linting fixed. App authentication works as intended. User is saved in the cache and log in to firebase is also persisted. Auth guard prevents unauthorised access.

File: src/screens/SignUp/Option3.jsx [hunk 1]
```diff
@@ -0,0 +1,262 @@
+import "react-native-gesture-handler";
```
Comment (samarth52) [inline src/screens/SignUp/Option3.jsx]: To clean up our codebase, delete `Option1.jsx`, `Option2.jsx`, and `SignUp.jsx`. Change `Option3.jsx` to `SignUp.tsx`. Remove unused variables and imports from this file. Add types wherever required.

File: src/screens/SignUp/Option3.jsx [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  currentStepIndicatorLabelFontSize: 15,
+  stepIndicatorLabelCurrentColor: "#000000",
+  stepIndicatorLabelFinishedColor: "#ffffff",
+  stepIndicatorLabelUnfinishedColor: "#000000",
+};
+
+const logo = require("../../assets/bei.jpg");
+
+//  Home Screen Navigation
+function SignUpOption3({ navigation }) {
```
Comment (samarth52) [inline src/screens/SignUp/Option3.jsx]: Add a type for `navigation` similar to screens like `MathMain`, but with reference to the AuthGuard stack.

File: src/screens/SignUp/Option3.jsx [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+              .catch((err) => {
+                setError("Email is already in use");
+                if (err.code === "auth/email-already-in-use") {
+                  setError("Email is already in use");
+                } else if (err.code === "auth/weak-password") {
+                  setError("Password is too short");
+                } else {
+                  setError("Unexpected error occured. Check your info");
+                }
+                console.log("error popped", error.code)
```
Comment (samarth52) [inline src/screens/SignUp/Option3.jsx]: This should be `err.code` instead of `error.code`

File: src/screens/SignUp/Option3.jsx [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+          <Text style={{ fontSize: 14, color: "#005AA3", fontWeight: "bold" }}>
+            Log In
+          </Text>
+        </Pressable>
+      </View>
+    </View>
+  );
+}
+
+SignUpOption3.propTypes = {
```
Comment (samarth52) [inline src/screens/SignUp/Option3.jsx]: Remove this when converting to Typescript

File: src/screens/SignUp/Option3.jsx [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+            height: 0.13 * Dimensions.get("window").width,
+          }}
+          titleStyle={styles.buttonTitle}
+          disabled={!isFormValid()}
+          title="Sign Up"
+          onPress={() => {
+            setError("");
+            emailSignUp(email, password)
+              .then((res) => {
+                console.log(typeof (res));
```
Comment (samarth52) [inline src/screens/SignUp/Option3.jsx]: Remove this line.

File: src/firebase/config.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+export default function firebaseInit() {
+  console.log("Initialising firebase");
+  const firebaseConfig: { [key: string]: string } = {
+    apiKey: "YOUR_KEY_HERE_AIzaSyAOWH",
+    authDomain: "your-auth-domain-b1234.firebaseapp.com",
+    databaseURL: "https://your-database-name.firebaseio.com",
+    projectId: "your-project-id-1234",
+    storageBucket: "your-project-id-1234.appspot.com",
+    messagingSenderId: "12345-insert-yourse",
+    appId: "insert yours: 1:1234:web:ee873bd1234c0deb7eba61ce",
```
Comment (samarth52) [inline src/firebase/config.ts]: Replace these with the actual config values from our project.

File: src/firebase/config.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { initializeApp } from "firebase/app";
+import { initializeAuth, getReactNativePersistence } from "firebase/auth";
+import ReactNativeAsyncStorage from "@react-native-async-storage/async-storage";
+
+// Need to be updated once google login screen is finalised
+export default function firebaseInit() {
+  console.log("Initialising firebase");
```
Comment (samarth52) [inline src/firebase/config.ts]: Remove this line

File: src/components/Auth/LogoutButton.tsx [hunk 8]
```diff
@@ -9,7 +9,7 @@ import * as Auth0 from "../../constants/Auth0";
 import { logout } from "../../redux/reducers/authReducer";
 import { setLoading } from "../../redux/reducers/loadingReducer/index";
 import Button from "../Button";
-import { useAuth0, Auth0Provider } from "react-native-auth0";
+import {getAuth, signOut} from "firebase/auth"
```
Comment (samarth52) [inline src/components/Auth/LogoutButton.tsx]: Remove lines 1, 7, 10, 14-15, and 45-63 (related to old OAuth code).

File: src/firebase/email_signin.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    auth,
+    email,
+    password,
+  );
+  // Signed in
+  const { user } = userCredential;
+  return user;
+  // ...
+}
+async function emailSignIn(email, password): Promise<User> {
```
Comment (samarth52) [inline src/firebase/email_signin.ts]: Add a string type to email and password parameters of these functions

File: src/screens/SignIn/SignIn.tsx [hunk 10]
```diff
@@ -0,0 +1,225 @@
+/* eslint-disable react-native-a11y/has-valid-accessibility-descriptors */
```
Comment (samarth52) [inline src/screens/SignIn/SignIn.tsx]: Remove this line

File: src/screens/SignIn/SignIn.tsx [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+    color: "white",
+  },
+  squareButtonTitle: {
+    alignSelf: "center",
+    fontSize: 20,
+    fontWeight: "400",
+  },
+  image: {
+    width: 100,
+    height: Platform.isPad ? 200 : Dimensions.get("window").height * 0.1,
```
Comment (samarth52) [inline src/screens/SignIn/SignIn.tsx]: Do `Platform.OS === "ios" && Platform.isPad` to pass the eslint checks

File: src/screens/SignIn/SignIn.tsx [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+          disabled={!isFormValid()}
+          title="Sign In"
+          onPress={() => {
+            setError("");
+            emailSignIn(email, password)
+              .then((res) => {
+                // !! Should add call to backend to retrieve rest of the information !!
+
+                const userObject: AuthUser = {
+                  _id: res.uid,
```
Comment (samarth52) [inline src/screens/SignIn/SignIn.tsx]: This `_id` should be used for MongoDB's id instead of Firebase's

File: src/screens/SignIn/SignIn.tsx [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+          <Text style={{ fontSize: 14, color: "#005AA3", fontWeight: "bold" }}>
+            Sign Up
+          </Text>
+        </Pressable>
+      </View>
+    </View>
+  );
+}
+
+SignInScreen.propTypes = {
```
Comment (samarth52) [inline src/screens/SignIn/SignIn.tsx]: Remove this

File: src/screens/SignIn/SignIn.tsx [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+  errorTitle: {
+    color: "#ed0707",
+    fontSize: 14,
+  },
+});
+
+const logo = require("../../assets/bei.jpg");
+
+//  Home Screen Navigation
+function SignInScreen({ navigation }) {
```
Comment (samarth52) [inline src/screens/SignIn/SignIn.tsx]: Add a type for `navigation` similar to how it's done in screens like `MathMain`, but with reference to the `AuthGuard` stack.

**Review by samarth52 — CHANGES_REQUESTED** [review]
Amazing job with this PR! There are a couple changes/refactoring to be made that I have mentioned in my comments. Create `AuthGuardStackParamList` similar to `RootStackParamList` so we can define the inputParams for the Login and Signup screens, and create a navigator called `AuthStack` using `AuthGuardStackParamList`, similar to the one created in `StackNavigator.tsx`. Great work!

File: src/components/Auth/LogoutButton.tsx (hunk 8 above)
Comment (thanasis457) [reply src/components/Auth/LogoutButton.tsx]: Removed Auth0 code and unnecessary imports.

File: src/screens/SignIn/SignIn.tsx (hunk 11 above)
Comment (thanasis457) [reply src/screens/SignIn/SignIn.tsx]: Not sure why this works but it does

File: src/screens/SignIn/SignIn.tsx (hunk 10 above)
Comment (thanasis457) [reply src/screens/SignIn/SignIn.tsx]: Removed and added necessary props to input fields. Also made the password field hide the text written to it.

File: src/screens/SignIn/SignIn.tsx (hunk 12 above)
Comment (thanasis457) [reply src/screens/SignIn/SignIn.tsx]: Got it. I assumed we would be using the same ids between Mongo and Firebase and Firebase would be the one providing the ids.
Anyway, I changed _id to idToken

File: src/screens/SignIn/SignIn.tsx (hunk 14 above)
Comment (thanasis457) [reply src/screens/SignIn/SignIn.tsx]: Take a look just to make sure I am not missing something. I looked through the React Navigation docs but I am still a bit cautious with my code.

File: src/screens/SignUp/Option3.jsx (hunk 2 above)
Comment (thanasis457) [reply src/screens/SignUp/Option3.jsx]: Also added PersonalInfoScreen to the routing props

**Comment by thanasis457** [comment]
I rebased by accident. The refactoring comment should include all the changes.

**Review by samarth52 — APPROVED** [review]
lgtm 👍

**Pull request merged**